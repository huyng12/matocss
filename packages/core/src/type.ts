export interface CSSProperty {
  prop: string;
  value: string;
}

export interface CSSRule {
  selector: string;
  properties: CSSProperty[];
}

export type ParsedCSS = CSSRule[];
