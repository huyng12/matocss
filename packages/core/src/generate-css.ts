import { CSSProperty } from "./type";

type GenerateCSSForAtomicRuleFunc = (
  className: string,
  cssProperty: CSSProperty
) => string;

export const generateCSSForAtomicRule: GenerateCSSForAtomicRuleFunc = (
  className,
  cssProperty
) => {
  return `:global(${className}) {
    ${cssProperty.prop}: ${cssProperty.value};
  }`;
};

type GenerateNewCSSForOldRuleFunc = (
  selector: string,
  classNames: string[]
) => string;

export const generateNewCSSForOldRule: GenerateNewCSSForOldRuleFunc = (
  selector,
  classNames
) => {
  return `${selector} {
    composes: ${classNames.join(" ")};
  }`;
};
