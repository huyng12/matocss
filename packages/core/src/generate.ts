import { CSSProperty } from "./type";

type GenerateCSSForAtomicRuleFunc = (
  className: string,
  cssProperty: CSSProperty
) => string;

export const generateCSSForAtomicRule: GenerateCSSForAtomicRuleFunc = (
  className,
  cssProperty
) => {
  const normalizedClassName = className.replaceAll(/[\[\]]/g, "\\$&");
  return `:global(.${normalizedClassName}) {
    ${cssProperty.prop}: ${cssProperty.value};
  }`;
};

type GenerateNewCSSForOldRuleFunc = (
  selector: string,
  classNames: string[]
) => string;

export const generateNewCSSForOldRule: GenerateNewCSSForOldRuleFunc = (
  selector,
  classNames
) => {
  return `${selector} {
    composes: ${classNames.join(" ")} from global;
  }`;
};
