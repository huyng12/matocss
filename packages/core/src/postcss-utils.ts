import { ChildNode, Rule, Declaration } from "postcss";

export const isRule = (node: ChildNode): node is Rule => node.type === "rule";

export const isDeclaration = (node: ChildNode): node is Declaration =>
  node.type === "decl";
