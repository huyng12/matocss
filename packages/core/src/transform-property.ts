type TransformPropertyFunc = (prop: string, value: string) => string;

// Transform CSS property to Atomic CSS name
// Example: `background-color: red` will be transformed to `bg-red`
export const transformProperty: TransformPropertyFunc = (prop, value) => {
  // TODO: Implement this function
  return "";
};
