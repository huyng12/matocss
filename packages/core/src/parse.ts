import { parse, Root } from "postcss";
import { isDeclaration, isRule } from "./postcss-utils";
import { CSSProperty, ParsedCSS } from "./type";

type NormalizePostCSSRootFunc = (root: Root) => ParsedCSS;

const normalizePostCSSRoot: NormalizePostCSSRootFunc = (root) => {
  return root.nodes.filter(isRule).map((node) => ({
    selector: node.selector,
    properties: node.nodes
      .filter(isDeclaration)
      .map((node): CSSProperty => ({ prop: node.prop, value: node.value })),
  }));
};

type ParseCSSFunc = (source: string) => ParsedCSS;

export const parseCSS: ParseCSSFunc = (source) => {
  const root = parse(source);
  return normalizePostCSSRoot(root);
};
