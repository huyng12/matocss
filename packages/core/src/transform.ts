import { parse } from "postcss";
import { generateNewCSSForOldRule } from "./generate-css";
import { isDeclaration, isRule } from "./postcss-utils";
import { transformProperty } from "./transform-property";

interface Params {
  source: string;
  generatedAtomicClassNames: Set<string>;
}

type TransformFunc = (params: Params) => string;

export const transform: TransformFunc = ({
  source,
  generatedAtomicClassNames,
}) => {
  const root = parse(source);

  const code: string[] = [];

  const rules = root.nodes.filter(isRule);
  for (const rule of rules) {
    const selector = rule.selector;
    const classNames: string[] = [];

    const decls = rule.nodes.filter(isDeclaration);
    for (const decl of decls) {
      const className = transformProperty(decl.prop, decl.value);
      classNames.push(className);
    }

    const css = generateNewCSSForOldRule(selector, classNames);
    code.push(css);
  }

  return code.join("\n");
};
