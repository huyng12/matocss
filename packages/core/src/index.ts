import { transform } from "./transform";

const generatedAtomicClassNames: Set<string> = new Set();

export const process = (source: string): string => {
  return transform({ source, generatedAtomicClassNames });
};
