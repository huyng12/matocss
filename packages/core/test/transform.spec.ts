import { test, expect } from "vitest";
import { transform } from "../src/transform";

test("transform css modules to enable atomic", () => {
  const generatedAtomicClassNames = new Set<string>();
  const source = `
    .container {
      color: white;
      background-color: red;
    }
  `;
  const css = transform({ source, generatedAtomicClassNames });
  expect(css).toMatchSnapshot();
});
