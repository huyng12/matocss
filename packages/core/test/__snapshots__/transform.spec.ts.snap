// Vitest Snapshot v1

exports[`transform css modules to enable atomic 1`] = `
":global(.color-\\\\[white\\\\]) {
    color: white;
  }
:global(.background-color-\\\\[red\\\\]) {
    background-color: red;
  }
.container {
    composes: color-[white] background-color-[red] from global;
  }"
`;
